rfcs/2021-01-08-5843-encdoing-decoding-format-vector.md

```diff
@@ -0,0 +1,95 @@
+
+It might not be totally clear from the serialization benchmark examples but while protobufs are inherently slow in some respects, they're _still_ more performant than raw JSON implementations and even MssagePack (which I would be willing to bet remains true for CBOR). Any schema based data format is likely going to have better perf than something schemaless. The schema itself is obviously part of the tooling overhead here but I'd argue that if we cared _more_ about tooling overhead than we did about perfomance and sustainability we'd probably be writing this project in something other than Rust.
+
+The truth is that of the formats that have wide adoption, all of them have their pitfalls and while protobufs bring with it some real tangible maintenance burden its still some of the _best_ tools with better integrations and libraries than most of the other things out there.
+
+As an intentionally buried lede - I wrote this in the hopes that folks would share their opinions on the subject so you might be able to find some gaps in theorem here.
+
+## Prior Art
+
+As mentioned above - my suggestion for the  data spec itself is that we _do_ in fact follow along with prior art. Should we follow this path, we should be giving `Open Telemetry` a thorough exploration. Per the format itself - I haven't found any single conglomerate conlusion about  _any_ encoding format. There are arguments for and against any that we could adopt.
+
+
+## Drawbacks
+
+The drawbacks to fully adopting protobufs on top of an OpenTelemetry spec are many, which makes the path no different than any other.
+    - Protobufs are slower than other schema'd data formats
+    - Protobufs have an unignorable maintenance and tooling overhead
+    - Protobufs are a heavy-weight solution to what could be percieved as a lightweight problem
+    - OpenTelemetry's data spec could be dramatically different than the way we structure our data today.
+
+Without question the existing burden of protobufs continues to exist if we follow this path.
+
+## Alternatives
+
+One alternative as mentioned already is flatbuffers and I didn't suggest this as a primary course of action for a few reasons. Like a lot of alternative encoding formats out there I'm honestly just not all that aware of what level of adoption it has. Also, While it has Rust support it seems rather lightweight and the level of integration available might not meet our needs. That being said, flatbuffers is really one of many many alternatives. Raw JSON with serde_bytes is another way to attack the problem. The alternatives are practically bottomless and to be perfectly honest, pretty subjective so it felt like the right approach here was to acknowledge that and submit an RFC to have a more structured conversation about those alternatives.
+
+## Outstanding Questions
+
+Should we abandon this proposal and instead do something entirely different?
+
+Should we continue on with protobufs and solve for our data structure leakage?
```

Comment:
It's worth being a little more specific about the "leakage" problem. In one sense, this is unavoidable since we're open source and sending this data our across the network. The things that we really want to avoid are:

1. Positioning this in any way as a new, public o11y data format
2. Coupling the serialized format so closely to our Rust code that we inadvertently change it (this is a potential downside of serde-based solutions)
3. Spending too much effort reinventing a wheel